# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: media/webrtc/signaling/src/peerconnection/PeerConnectionImpl.cpp
# Commit: c54cb3c10992
# Full Hash: c54cb3c109922bc15fb6d0ca67e5f2a05980c1a5
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2019-07-31 15:46:43
# Regressor Bug: 1493613
# File Overlap Count: 2
# Description:
#   Bug 1493613 - Move MediaStream control from DOMMediaStream to MediaStreamTrack. r=padenot
#   
#   This is inherently large, because modifying these bits of DOMMediaStream and
#   MediaStreamTrack affects all consumers and producers of all DOMMediaStreams and
#   MediaStreamTracks.
# ==============================================================================

diff -r 0b03dd9d20ac -r c54cb3c10992 media/webrtc/signaling/src/peerconnection/PeerConnectionImpl.cpp
--- a/media/webrtc/signaling/src/peerconnection/PeerConnectionImpl.cpp	Wed Jul 31 09:35:06 2019 +0000
+++ b/media/webrtc/signaling/src/peerconnection/PeerConnectionImpl.cpp	Wed Jul 31 07:58:17 2019 +0000
@@ -35,6 +35,7 @@
 
 #include "AudioConduit.h"
 #include "VideoConduit.h"
+#include "MediaStreamGraph.h"
 #include "runnable_utils.h"
 #include "PeerConnectionCtx.h"
 #include "PeerConnectionImpl.h"
@@ -1797,18 +1798,7 @@
     SdpMediaSection::MediaType type) {
   bool audio = (type == SdpMediaSection::MediaType::kAudio);
 
-  MediaStreamGraph* graph = MediaStreamGraph::GetInstance(
-      audio ? MediaStreamGraph::AUDIO_THREAD_DRIVER
-            : MediaStreamGraph::SYSTEM_THREAD_DRIVER,
-      GetWindow(), MediaStreamGraph::REQUEST_DEFAULT_SAMPLE_RATE);
-
-  RefPtr<DOMMediaStream> stream =
-      DOMMediaStream::CreateSourceStreamAsInput(GetWindow(), graph);
-
-  CSFLogDebug(LOGTAG, "Created media stream %p, inner: %p", stream.get(),
-              stream->GetInputStream());
-
-  // Set the principal used for creating the tracks. This makes the stream
+  // Set the principal used for creating the tracks. This makes the track
   // data (audio/video samples) accessible to the receiving page. We're
   // only certain that privacy hasn't been requested if we're connected.
   nsCOMPtr<nsIPrincipal> principal;
@@ -1817,29 +1807,39 @@
   if (mPrivacyRequested.isSome() && !*mPrivacyRequested) {
     principal = doc->NodePrincipal();
   } else {
-    // we're either certain that we need isolation for the streams, OR
-    // we're not sure and we can fix the stream in SetDtlsConnected
+    // we're either certain that we need isolation for the tracks, OR
+    // we're not sure and we can fix the track in SetDtlsConnected
     principal =
         NullPrincipal::CreateWithInheritedAttributes(doc->NodePrincipal());
   }
 
+  MediaStreamGraph* graph = MediaStreamGraph::GetInstance(
+      audio ? MediaStreamGraph::AUDIO_THREAD_DRIVER
+            : MediaStreamGraph::SYSTEM_THREAD_DRIVER,
+      GetWindow(), MediaStreamGraph::REQUEST_DEFAULT_SAMPLE_RATE);
+
   RefPtr<MediaStreamTrack> track;
   RefPtr<RemoteTrackSource> trackSource;
+  RefPtr<SourceMediaStream> source = graph->CreateSourceStream();
   if (audio) {
-    trackSource = new RemoteTrackSource(principal,
+    trackSource = new RemoteTrackSource(source, principal,
                                         NS_ConvertASCIItoUTF16("remote audio"));
-    track = stream->CreateDOMTrack(333,  // Use a constant TrackID. Dependents
-                                         // read this from the DOM track.
-                                   MediaSegment::AUDIO, trackSource);
+    track = new AudioStreamTrack(GetWindow(), source,
+                                 333,  // Use a constant TrackID. Dependents
+                                       // read this from the DOM track.
+                                 trackSource);
   } else {
-    trackSource = new RemoteTrackSource(principal,
+    trackSource = new RemoteTrackSource(source, principal,
                                         NS_ConvertASCIItoUTF16("remote video"));
-    track = stream->CreateDOMTrack(666,  // Use a constant TrackID. Dependents
-                                         // read this from the DOM track.
-                                   MediaSegment::VIDEO, trackSource);
+    track = new VideoStreamTrack(GetWindow(), source,
+                                 666,  // Use a constant TrackID. Dependents
+                                       // read this from the DOM track.
+                                 trackSource);
   }
 
-  stream->AddTrackInternal(track);
+  CSFLogDebug(LOGTAG, "Created %s track %p, inner: %p",
+              audio ? "audio" : "video", track.get(), track->GetStream());
+
   // Spec says remote tracks start out muted.
   trackSource->SetMuted(true);
 